src/server/pps/server/api_server.go

```diff
@@ -1477,185 +1502,157 @@ func (a *apiServer) jobManager(ctx context.Context, jobInfo *pps.JobInfo) {
 			// a deployment is idempotent: nothing happens if the workers have
 			// already been scaled.
-			deploymentName := PipelineDeploymentName(jobInfo.Pipeline.Name, jobInfo.PipelineVersion)
+			deploymentName = PipelineDeploymentName(jobInfo.Pipeline.Name, jobInfo.PipelineVersion)
 			if err := a.scaleUpWorkers(ctx, deploymentName, jobInfo.ParallelismSpec); err != nil {
 				return err
 			}
-			wp, err = a.newWorkerPool(ctx, deploymentName, jobInfo.Job.ID)
-			if err != nil {
-				return err
-			}
 		} else {
-			wp, err = a.newWorkerPool(ctx, JobDeploymentName(jobInfo.Job.ID), jobInfo.Job.ID)
-			if err != nil {
-				return err
-			}
+			deploymentName = JobDeploymentName(jobInfo.Job.ID)
 		}
 
-		// We have a goroutine that receives the datums that fail to
-		// be processed, and put them back onto the datum queue.
-		jobFailedCh := make(chan struct{})
-		go func() {
-			var dts []*datum
-			for {
-				if len(dts) > 0 {
-					select {
-					case wp.DataCh() <- dts[0]:
-						protolion.Infof("retrying datum %v", dts[0].files)
-						dts = dts[1:]
-					case dt := <-wp.FailCh():
-						if dt.retries >= MaximumRetriesPerDatum {
-							close(jobFailedCh)
-							return
-						}
-						protolion.Infof("datum %v is queued up for retry", dt.files)
-						dts = append(dts, dt)
-					case <-ctx.Done():
-						return
-					}
-				} else {
-					select {
-					case dt := <-wp.FailCh():
-						if dt.retries >= MaximumRetriesPerDatum {
-							close(jobFailedCh)
-							return
-						}
-						protolion.Infof("datum %v is queued up for retry", dt.files)
-						dts = append(dts, dt)
-					case <-ctx.Done():
-						return
-					}
-				}
-			}
-		}()
-
+		failed := false
+		numWorkers, err := a.numWorkers(ctx, deploymentName)
+		if err != nil {
+			return err
+		}
+		limiter := limit.New(numWorkers)
 		// process all datums
 		df, err := newDatumFactory(ctx, pfsClient, jobInfo.Inputs, nil)
 		if err != nil {
 			return err
 		}
-		var inflightData int64
-		var processedData int64
-		var etcdProcessedData int64 // the value of processedData we've sent to etcd
+		tree := hashtree.NewHashTree()
+		var treeMu sync.Mutex
+
+		processedData := int64(0)
+		setProcessedData := int64(0)
 		totalData := int64(df.Len())
-		// This goro is responsible for updating job progress
-		// The values sent on this channel will be the number of datums
-		// processed so far.
-		progressCh := make(chan int64)
-		go func() {
-			var processed int64
-			for {
-				// We do the first update before receiving a progress,
-				// because we want to update the total number of datums.
+		var progressMu sync.Mutex
+		updateProgress := func(processed int64) {
+			progressMu.Lock()
+			defer progressMu.Unlock()
+			processedData += processed
+			// so as not to overwhelm etcd we update at most 100 times per job
+			if (float64(processedData-setProcessedData)/float64(totalData)) > .01 ||
+				processedData == 0 || processedData == totalData {
```

Comment:
I set it below only after the value has actually been set in etcd. This means that if for some reason a value fails to set we'll try again on the next one. I think this makes sense, although maybe it doesn't because if we get a failure from etcd it might be because it's under too much load and this would just overwhelm it.